README.md
# libimlayouts

debian/changelog
libimlayouts (1.6.3) unsatable; urgency=low

  * port to debhelper 20

 -- Ivaylo Dimitrov <ivo.g.dimitrov.75@gmail.com>  Sun, 09 Feb 2020 14:52:33 +0200

libimlayouts (1.6.2) unsatable; urgency=low

  * Fix GCC complaining

 -- Merlijn Wajer <merlijn@wizzup.org>  Fri, 24 Jan 2020 15:10:05 +0100

libimlayouts (1.6.1) unsatable; urgency=low

  * build platform-specific -dev package

 -- Ivaylo Dimitrov <ivo.g.dimitrov.75@gmail.com>  Tue, 23 Aug 2018 14:48:21 +0300

libimlayouts (1.6.0) unsatable; urgency=low

  * add vkb manipulation programs to -dev package

 -- Ivaylo Dimitrov <ivo.g.dimitrov.75@gmail.com>  Mon, 20 Aug 2018 20:22:43 +0300

libimlayouts (1.5.17) unsatable; urgency=low

  * fix a segfault if layout creation has failed
  * do not try to fread() zero-sized blocks

 -- Ivaylo Dimitrov <ivo.g.dimitrov.75@gmail.com>  Mon, 11 Jun 2018 16:40:23 +0300

libimlayouts (1.5.16) unsatable; urgency=low

  * Finish RE of functions needed so far

 -- Ivaylo Dimitrov <ivo.g.dimitrov.75@gmail.com>  Sat, 19 May 2018 09:39:41 +0300

libimlayouts (1.5.15) unsatable; urgency=low

  * Some cleanups

 -- Jonathan Wilson <jfwfreo@tpgi.com.au>  Sat, 28 Oct 2017 08:08:40 -0400

libimlayouts (1.5.14) unsatable; urgency=low

  * Initial RE release.

 -- Jonathan Wilson <jfwfreo@tpgi.com.au>  Fri, 27 Oct 2017 21:49:40 -0400


imlayouts.h
/**
  @file imlayout.h

  Copyright (C) 2011 Jonathan Wilson

  @author Jonathan wilson <jfwfreo@tpgi.com.au>

  This library is free software; you can redistribute it and/or modify it
  under the terms of the GNU Lesser General Public License version 2.1 as
  published by the Free Software Foundation.

  This library is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser
  General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with this library; if not, write to the Free Software Foundation,
  Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*/
typedef struct _vkb_key vkb_key;
struct _vkb_key
{
	unsigned char gtk_state;
	unsigned char unk1;
	unsigned short key_flags;
	unsigned char current_slide_key;
	unsigned int width;
	unsigned char byte_count;
	gchar **labels;
	unsigned int special_font;
	unsigned char key_type;
	unsigned int left;
	unsigned int top;
	unsigned int right;
	unsigned int bottom;
	unsigned int unk9;
	unsigned char offset;
	unsigned char scancode_length;
	unsigned char unk12;
	unsigned char unk13;
	unsigned char *scancode;
	unsigned char num_sub_keys;
	vkb_key *sub_keys;
	unsigned char key_size;
};
#define KEY_TYPE_NORMAL 0
#define KEY_TYPE_SLIDING 1
#define KEY_TYPE_MODIFIER 2
#define KEY_TYPE_MULTIPLE 4
#define KEY_TYPE_ALPHA          0x01
#define KEY_TYPE_NUMERIC        0x02
#define KEY_TYPE_HEXA           0x04
#define KEY_TYPE_TELE           0x08
#define KEY_TYPE_SPECIAL        0x10
#define KEY_TYPE_DEAD           0x20
#define KEY_TYPE_WHITESPACE     0x40
#define KEY_TYPE_EXTRA_BYTE     0x80
#define KEY_TYPE_RAW            ((0x01<<8) | KEY_TYPE_EXTRA_BYTE)
#define KEY_TYPE_TONE           ((0x02<<8) | KEY_TYPE_EXTRA_BYTE)
#define KEY_TYPE_TAB            ((0x04<<8) | KEY_TYPE_EXTRA_BYTE)
#define KEY_TYPE_BACKSPACE      ((0x08<<8) | KEY_TYPE_EXTRA_BYTE)
#define KEY_TYPE_SHIFT          ((0x10<<8) | KEY_TYPE_EXTRA_BYTE)

typedef struct _vkb_key_section
{
	unsigned char num_keys;
	unsigned char num_rows;
	unsigned char *num_keys_in_rows;
	vkb_key *keys;
	unsigned char margin_left;
	unsigned char margin_top;
	unsigned char margin_bottom;
	unsigned char margin_right;
	unsigned int width;
	unsigned int unk0;
} vkb_key_section;

typedef struct _vkb_sub_layout
{
	unsigned char type;
	unsigned char variance_index;
	unsigned char num_key_sections;
	vkb_key_section *key_sections;
	gchar *label;
	unsigned int hash;
	unsigned int width;
	unsigned int height;
} vkb_sub_layout;
#define SUBLAYOUT_TYPE_LOWERCASE 0
#define SUBLAYOUT_TYPE_UPPERCASE 1
#define SUBLAYOUT_TYPE_SPECIAL 2

typedef struct _vkb_layout
{
	unsigned char type;
	unsigned char numeric;
	unsigned char num_sub_layouts;
	vkb_sub_layout *sub_layouts;
	unsigned char default_key_size;
} vkb_layout;
#define LAYOUT_TYPE_NORMAL 0
#define LAYOUT_TYPE_SPECIAL 1
#define LAYOUT_TYPE_LARGE 3
#define LAYOUT_TYPE_THUMB 4
#define LAYOUT_TYPE_PINYIN 5
#define LAYOUT_TYPE_ZHUYIN 6
#define LAYOUT_TYPE_CANGJIE 7
#define LAYOUT_TYPE_STROKE 8

typedef struct _vkb_key_size
{
	unsigned char width;
	unsigned char height;
	unsigned char baseline;
	unsigned char margin_left;
	unsigned char margin_top;
} vkb_key_size;

typedef struct _vkb_layout_collection
{
	char version;
	unsigned char num_layouts;
	char *name;
	char *lang;
	char *wc;
	char *filename;
	int *layout_types;
	int special_font;
	unsigned short *offsets;
	unsigned char num_key_sizes;
	vkb_key_size *key_sizes;
	unsigned char num_screen_modes;
	unsigned char *screen_modes;
} vkb_layout_collection;
#define SCREEN_TYPE_NORMAL 0
#define SCREEN_TYPE_THUMB 1
#define SCREEN_TYPE_ROTATED 2
typedef struct _vkb_keyboard_layout {
	int header_fd;
	int body_fd;
	vkb_layout_collection layout;
	vkb_layout *layouts;
	char *file;
	char *header_file;
	char *body_file;
	int is_key;
	int is_slide_key;
	int is_multiple_key;
	int parsing_multiple;
} vkb_keyboard_layout;
void imlayout_vkb_free_layout(vkb_layout *layout);
void imlayout_vkb_free_layout_collection(vkb_layout_collection *collection);
void imlayout_vkb_free_layout_list(GSList *list);
vkb_layout *imlayout_vkb_get_layout(vkb_layout_collection *collection, int layout_type);
GSList *imlayout_vkb_get_layout_list();
GSList *imlayout_vkb_get_layout_list_from_prefix(const gchar *path);
int imlayout_vkb_init_buttons(vkb_layout_collection *collection, vkb_layout *section, int width, int height);
int vkb_init_buttons(vkb_layout_collection *collection, vkb_layout *section);
vkb_layout_collection *imlayout_vkb_load_file(const char *fname);
void print_info(vkb_layout_collection *collection);
void print_layout_info(vkb_layout *layout);
void print_sublayout_info(vkb_sub_layout *layout);
void add_key(vkb_keyboard_layout *layout);
void add_keysection(vkb_keyboard_layout *layout_name);
void add_keysize(vkb_keyboard_layout *layout);
void add_layout(vkb_keyboard_layout *layout);
void add_rows(vkb_keyboard_layout *layout);
void add_screen(vkb_keyboard_layout *layout);
void add_slide(vkb_keyboard_layout *layout, const char *src, int len);
void add_sublayout(vkb_keyboard_layout *layout);
vkb_key *get_key(vkb_keyboard_layout *layout);
vkb_key *get_key0(vkb_keyboard_layout *layout, int no_subkey);
vkb_key_section *get_keysection(vkb_keyboard_layout *layout);
vkb_key_size *get_keysize(vkb_keyboard_layout *layout);
vkb_layout *get_layout(vkb_keyboard_layout *layout);
unsigned char *get_num_keys_in_row(vkb_keyboard_layout *layout);
unsigned char *get_screen(vkb_keyboard_layout *layout);
vkb_sub_layout *get_sublayout(vkb_keyboard_layout *layout);
void increase_num_keys_in_row(vkb_keyboard_layout *layout);


decode_vkb.pl
#!/usr/bin/perl -w

# decode_vkb is a perl script to decode binary virtual keyboard definition
# files from Nokia n800 Internet tablet to a XML format for editing them.
# The generated XML-files can be encoded back to binary .vkb format using
# gen_vkb utility from Nokia.
#
# (C) 2007 Kimmo Jukarainen <kimju-2007@inside.org>
#
# This program is licensed under GPLv2. See LICENSE.txt

# -------------------------------------------------------------------- #

# usage: ./decode_vkb file.vkb > file.xml

# -------------------------------------------------------------------- #

# Version: 0.01 - 2007-08-21
#  - initial release

# -------------------------------------------------------------------- #

use strict;
use vars qw/$fh $idt/; # the ugly globals: filehandle and indentation level

# -------------------------------------------------------------------- #

die "Invalid parameters.\n" .
    "Usage: $0 file.vkb > file.xml\n" unless scalar @ARGV == 1;
die "Can't open file\n" unless open($fh, $ARGV[0]);
binmode $fh;

# -------------------------------------------------------------------- #

sub getByteInt() {
  my $tmp;
  read($fh, $tmp, 1);
  return ord($tmp);
}

sub getWordInt() {
  my $tmp = getByteInt;
  $tmp += getByteInt << 8;
  return $tmp;
}

sub getString() {
  my $len = getByteInt();
  my $tmp;
  read($fh, $tmp, $len);
  return $tmp;
}

sub getStringBytes() {
  my $len = getByteInt();
  my $tmp;
  for (my $i = 0; $i < $len; $i++) {
    $tmp .= getByteInt();
    $tmp .= ' ';
  }
  return $tmp;
}
 

# -------------------------------------------------------------------- #

sub xprint($) {
  my $line = shift;
  print "\t" x $idt, $line, "\n";
}

# -------------------------------------------------------------------- #

sub escapeXML($) {
  my $str = shift;
  $str =~ s:&:&amp;:g;
  $str =~ s:<:&lt;:g;
  $str =~ s:>:&gt;:g;
  return $str;
}

# -------------------------------------------------------------------- #

sub decodeHeader() {
  # 00000000  01 02 0d 53 75 6f 6d 69  20 28 53 75 6f 6d 69 29  |...Suomi (Suomi)|
  # 00000010  05 66 69 5f 46 49 05 66  69 5f 46 49 02 00 01 05  |.fi_FI.fi_FI....|
  # 00000020  23 23 1a 00 00 32 23 1a  00 00 37 23 1a 00 00 48  |##...2#...7#...H|
  # 00000030  37 29 00 00 8f 37 29 00  00 51 00 58 02 00 00 00  |7)...7)..Q.X....|
  # 00000040  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|

  my %vkb;
  $vkb{'version'} = getByteInt();
  $vkb{'layouts'} = getByteInt();
  $vkb{'langname'} = getString();
  $vkb{'langcode'} = getString();
  $vkb{'wordcomp'} = getString();
  $vkb{'screenmodes'} = getByteInt();
  for (my $i = 0; $i < $vkb{'screenmodes'}; $i++) {
    $vkb{'screenmode_'.$i} = getByteInt();
  }
  $vkb{'keysizes'} = getByteInt();
  for (my $i = 0; $i < $vkb{'keysizes'}; $i++) {
    my %keysize;
    $keysize{'width'} = getByteInt();
    $keysize{'height'} = getByteInt();
    $keysize{'baseline'} = getByteInt();
    $keysize{'top'} = getByteInt();
    $keysize{'left'} = getByteInt();
    $vkb{'keysize_'.$i} = \%keysize;
  }

  # There's no information about how many layouts the file has.
  # I'm guessing that there can be as many layouts as 16-bit 
  # offsets can handle.
  # The vkb specification says that there are 20 reserved bytes
  # after the offset array. Currently those are all zeros, so I'm
  # using them as sentinel to the offset array and read offsets
  # until we hit the offset 0x0000.
  
  my $tmp;
  my $i = 0;
  while ($tmp = getWordInt()) {
    $vkb{'offset_'.$i} = $tmp;
    $i++;
  }
  $vkb{'offsets'} = $i;
  
  return %vkb;
}

# -------------------------------------------------------------------- #

sub printHeader($) {

  my %vkb = %{$_[0]};
  $idt = 0;

  my $fname = $ARGV[0];
  $fname =~ s:.*/([^/]+)$:$1:;
  $fname =~ s:\.([^.]+)$::;

  xprint 
    '<keyboards file="' . $fname . '" version="' . $vkb{'version'} . 
    '" lang="' . $vkb{'langcode'} . '" name="' . $vkb{'langname'} . 
    '" wc="' . $vkb{'wordcomp'} .'">';

  $idt++;
  xprint '<screen>';
  $idt++;

  my %screenmodes = ( '0' => 'NORMAL', '1' => 'FULL', '2' => 'ROTATED' );
  for (my $i = 0; $i < $vkb{'screenmodes'}; $i++) {
    if (defined $screenmodes{$vkb{'screenmode_'.$i}}) {
      xprint '<mode name="'.$screenmodes{$vkb{'screenmode_'.$i}}.'"/>';
    } else {
      xprint '<mode name="'.$vkb{'screenmode_'.$i}.'"/>';
      warn "Unknown screenmode: " . $vkb{'screenmode_'.$i};
    }
  }
  $idt--;
  xprint '</screen>';

  xprint "<keysizes>";
  $idt++;
  for (my $i = 0; $i < $vkb{'keysizes'}; $i++) {
    my %k = %{$vkb{'keysize_'.$i}};
    xprint
      '<size height="'.$k{'height'}.
      '" baseline="'.$k{'baseline'}.
      '" width="'.$k{'width'}.
      '" margin_top="'.$k{'top'}.
      '" margin_left="'.$k{'left'}.
      '" />';
  }
  $idt--;
  xprint "</keysizes>";

}

# -------------------------------------------------------------------- #

sub decodeLayout($) {

  my $offset=shift;
  seek($fh, $offset, 0);

  my %layout;

  $layout{'type'} = getByteInt();
  $layout{'subs'} = getByteInt();
  $layout{'numeric'} = getByteInt();
  $layout{'defsize'} = getByteInt();
  
  for (my $i = 0; $i < $layout{'subs'}; $i++) {
    $layout{'sub_'.$i} = decodeSubLayout();
  }
  
  return \%layout;
}

# -------------------------------------------------------------------- #

sub printLayout($) {
  my %l = %{$_[0]};

  my %layouttypes = ( 
    '0' => 'NORMAL',
    '1' => 'SPECIAL',
    '3' => 'LARGE',
    '4' => 'THUMB',
    '5' => 'PINYIN',
    '6' => 'ZHUYIN',
    '7' => 'CANGJIE',
    '8' => 'STROKE'
  );

  if (defined $layouttypes{$l{'type'}}) {
    xprint '<keyboard layout="'.$layouttypes{$l{'type'}}.
      '" default_key_size="'.$l{'defsize'}.'"' . 
      ( $l{'numeric'} ? ' numeric="'.$l{'numeric'}.'"' : '' ) . '>';
  } else {
    xprint '<keyboard layout="'.$l{'type'}.
      '" default_key_size="'.$l{'defsize'}.'"'.
      ( $l{'numeric'} ? ' numeric="'.$l{'numeric'}.'"' : '' ) . '>';

    warn "Unknown layouttype: " . $l{'type'};
  }

  $idt++;
  for (my $i = 0; $i < $l{'subs'}; $i++) {
    printSubLayout($l{'sub_'.$i});
  }
  $idt--;
  xprint '</keyboard>';
}

# -------------------------------------------------------------------- #

sub decodeSubLayout() {

  my %sublay;

  $sublay{'type'} = getByteInt();
  $sublay{'variance'} = getByteInt();
  $sublay{'label'} = getString();
  $sublay{'keysects'} = getByteInt();

  for (my $i = 0; $i < $sublay{'keysects'}; $i++) {
    $sublay{'keysect_'.$i} = decodeKeysect();
  }
  
  return \%sublay;
}

# -------------------------------------------------------------------- #

sub printSubLayout($) {
  my %s = %{$_[0]};

  my %sublayouttypes = (
    '0' => 'LOWERCASE',
    '1' => 'UPPERCASE',
    '2' => 'SINGLE',
  );
  
  if (defined $sublayouttypes{$s{'type'}}) {
    xprint '<sublayout type="'.$sublayouttypes{$s{'type'}}.
      '" variance_index="'.$s{'variance'}.'" label="'.$s{'label'}.'">';
  } else {
    xprint '<sublayout type="'.$s{'type'}.
      '" variance_index="'.$s{'variance'}.'" label="'.$s{'label'}.'">';
    warn "Unknown sublayouttype: " . $s{'type'};
  }
  
  $idt++;
  for (my $i = 0; $i < $s{'keysects'}; $i++) {
    printKeysect($s{'keysect_'.$i});
  }
  $idt--;
  xprint '</sublayout>';
}

# -------------------------------------------------------------------- #

sub decodeKeysect() {

  my %keysect;

  $keysect{'keys'} = getByteInt();
  $keysect{'rows'} = getByteInt();
  $keysect{'left'} = getByteInt();
  $keysect{'top'} = getByteInt();
  $keysect{'bottom'} = getByteInt();
  $keysect{'right'} = getByteInt();

  for (my $i = 0; $i < $keysect{'rows'}; $i++) {
    $keysect{'rowsize_'.$i} = getByteInt;
  }

  for (my $i = 0; $i < $keysect{'rows'}; $i++) {
    $keysect{'row_'.$i} = [];
    for (my $j = 0; $j < $keysect{'rowsize_'.$i}; $j++) {
      push(@{$keysect{'row_'.$i}}, decodeKey());
    }
  }

  return \%keysect;
}

# -------------------------------------------------------------------- #

sub printKeysect($) {
  my %k = %{$_[0]};
  
  xprint '<keysection margin_top="'.$k{'top'}.
    '" margin_left="'.$k{'left'}.
    '" margin_bottom="'.$k{'bottom'}.
    '" margin_right="'.$k{'right'}.'">';
  $idt++;
  for (my $i = 0; $i < $k{'rows'}; $i++) {
    printKeyrow($k{'row_'.$i});
  }
  $idt--;
  xprint '</keysection>';
}

# -------------------------------------------------------------------- #

sub printKeyrow($) {
  my @row = @{$_[0]};
  xprint '<row>';
  $idt++;
  
  foreach my $keyref (@row) {
    printKey($keyref);
  }
  
  $idt--;
  xprint '</row>';

}

# -------------------------------------------------------------------- #

sub decodeKey();
sub decodeKey() {
  # normal and slide:
  # 000002f0  00 01 01 76 03 00 05 01  62 03 00 01 01 6e 03 00  |...v....b....n..|
  # 00000300  01 01 6d 03 01 10 8a 01  2e 01 2c 01 2d 01 21 01  |..m.......,.-.!.|
  # 00000310  3f 01 26 01 3b 01 3a 01  5f 01 40 03 01 00 03 41  |?.&.;.:._.@....A|
  # 00000320  42 43 01 1f 03 00 00 00  00 0b 0b 09 00 01 01 51  |BC.............Q|

  my %key;
  $key{'type'} = getByteInt();
  $key{'special_font'} = 0;
  if ($key{'type'} & 0x80) {
    $key{'type'} = $key{'type'} ^ 0x80;
    $key{'special_font'} = 1;
  }
  $key{'count'} = 1;

  if ($key{'type'} == 4 ) {
    # multiple
    # 00000070  00 00 00 01 72 00 01 d2  04 07 00 01 01 74 00 00  |....r........t..|
    # 00000080  01 01 79 00 00 01 01 75  00 00 01 01 69 00 00 01  |..y....u....i...|
    # 00000090  01 6f 00 00 09 01 70 00  00 01 02 c3 a5 01 00 05  |.o....p.........|
    # 000000a0  01 61 00 00 01 01 73 00  00 05 01 64 00 00 05 01  |.a....s....d....|
    # type size # 1st normal key # 2nd            #     # last              #
    # 04   07   # 00 01 01 74 00 # 00 01 01 79 00 # ... # 00 01 02 c3 a5 01 # 

    $key{'count'} = getByteInt();

    my @tmpkeys;
    for (my $i = 0; $i < $key{'count'}; $i++) {
      push(@tmpkeys, decodeKey());
    }
    $key{'keys'} = \@tmpkeys;
  } else {
    $key{'attributes'} = getByteInt;
    if ($key{'attributes'} & 0x80) {
      $key{'attributes'} += (getByteInt() << 8);
    }
    if ($key{'type'} == 1) {
      $key{'count'} = getByteInt() ^ 0x80;
      my @labels;
      for (my $i = 0; $i < $key{'count'}; $i++ )
      {
        my $label = getString;
        push(@labels, $label)
      }
      $key{'labels'} = \@labels;
    } else {
      $key{'label'} = getString;
    }
    $key{'size'} = getByteInt();
    if ($key{'attributes'} & 0x100) {
      $key{'scancode'} = getStringBytes;
    }
  }
  return \%key;
}

# -------------------------------------------------------------------- #

sub printKey($);
sub printKey($) {
  my $keyref = shift;
  if ( $keyref->{'type'} == 0 || $keyref->{'type'} == 2 
     ) {
     # normal
     xprint '<key size="' . $keyref->{'size'} . '"' .
       (($keyref->{'special_font'} == 1) ? ' special_font="SPECIAL_FONT"' : "" ) .
       (($keyref->{'attributes'} & (0x01)) ? ' alpha="ALPHA"' : "" ) .
       (($keyref->{'attributes'} & (0x02)) ? ' numeric="NUMERIC"' : "" ) .
       (($keyref->{'attributes'} & (0x04)) ? ' hexa="HEXA"' : "" ) .
       (($keyref->{'attributes'} & (0x08)) ? ' tele="TELE"' : "" ) .
       (($keyref->{'attributes'} & (0x10)) ? ' special="SPECIAL"' : "" ) .
       (($keyref->{'attributes'} & (0x20)) ? ' dead="DEAD"' : "" ) .
       (($keyref->{'attributes'} & (0x40)) ? ' whitespace="WHITESPACE"' :"" ) .
       (($keyref->{'attributes'} & (0x100)) ? ' raw="RAW" scancode="' . $keyref->{'scancode'} . '"'  :"" ) .
       (($keyref->{'attributes'} & (0x200)) ? ' tone="TONE"' :"" ) .
       (($keyref->{'attributes'} & (0x400)) ? ' tab="TAB"' :"" ) .
       (($keyref->{'attributes'} & (0x800)) ? ' backspace="BACKSPACE"' :"" ) .
       (($keyref->{'attributes'} & (0x1000)) ? ' shift="SHIFT"' :"" ) .
       (($keyref->{'type'} == 0x02 ) ? ' modifier="MODIFIER"' : "" ) .
       '>' . escapeXML($keyref->{'label'}) . '</key>';

  } elsif ( $keyref->{'type'} == 1 ) {
    # slide
    xprint '<key type="SLIDE" size="' . $keyref->{'size'} . '"' .
      (($keyref->{'special_font'} == 1) ? ' special_font="SPECIAL_FONT"' : "" ) .
      (($keyref->{'attributes'} & (0x01)) ? ' alpha="ALPHA"' : "" ) .
      (($keyref->{'attributes'} & (0x02)) ? ' numeric="NUMERIC"' : "" ) .
      (($keyref->{'attributes'} & (0x04)) ? ' hexa="HEXA"' : "" ) .
      (($keyref->{'attributes'} & (0x08)) ? ' tele="TELE"' : "" ) .
      (($keyref->{'attributes'} & (0x10)) ? ' special="SPECIAL"' : "" ) .
      (($keyref->{'attributes'} & (0x20)) ? ' dead="DEAD"' : "" ) .
      (($keyref->{'attributes'} & (0x40)) ? ' whitespace="WHITESPACE"' :"" ) .
      (($keyref->{'attributes'} & (0x100)) ? ' raw="RAW" scancode="' . $keyref->{'scancode'} . '"'  :"" ) .
      (($keyref->{'attributes'} & (0x200)) ? ' tone="TONE"' :"" ) .
      (($keyref->{'attributes'} & (0x400)) ? ' tab="TAB"' :"" ) .
      (($keyref->{'attributes'} & (0x800)) ? ' backspace="BACKSPACE"' :"" ) .
      (($keyref->{'attributes'} & (0x1000)) ? ' shift="SHIFT"' :"" ) .
      '>';

    $idt++;
    foreach my $a (@{$keyref->{'labels'}}) {
      xprint '<slide>' . escapeXML($a) . '</slide>';
    }
    $idt--;
    xprint '</key>';

  } elsif ( $keyref->{'type'} == 4 ) {
    # multiple
    xprint '<multiple>';
    $idt++;
    for $a (@{$keyref->{keys}}) {
      printKey($a);
    }
    $idt--;
    xprint '</multiple>';
  } else {
    warn "Unknown key type"
  }
}

# -------------------------------------------------------------------- #

sub printXML($) {
  my $vkbref = shift;
  printHeader($vkbref);
  for (my $i = 0; $i < $vkbref->{'offsets'}; $i++) {
    printLayout($vkbref->{'layout_'.$i});
  }
  $idt--;
  xprint '</keyboards>';
}

# -------------------------------------------------------------------- #

my %vkb = decodeHeader;

for (my $i = 0; $i < $vkb{'offsets'}; $i++) {
  $vkb{'layout_'.$i} = decodeLayout($vkb{'offset_'.$i});
}

printXML(\%vkb);

# -------------------------------------------------------------------- #
